### ISSUE TITLE
importlib.metadata.version can return None

### ISSUE THREAD

**Issue by 572e78c3-960c-4f26-a09b-2a09f49b46e4** [python/cpython#91216]
BPO | [47060](https://bugs.python.org/issue47060)
--- | :---
Nosy | @jaraco, @DMRobertson

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/jaraco'
closed_at = None
created_at = <Date 2022-03-18.18:11:51.491>
labels = ['type-bug', 'library', '3.10']
title = 'importlib.metadata.version can return None'
updated_at = <Date 2022-03-20.19:58:49.043>
user = 'https://github.com/DMRobertson'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-20.19:58:49.043>
actor = 'jaraco'
assignee = 'jaraco'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2022-03-18.18:11:51.491>
creator = 'David Robertson'
dependencies = []
files = []
hgrepos = []
issue_num = 47060
keywords = []
message_count = 2.0
messages = ['415516', '415529']
nosy_count = 2.0
nosy_names = ['jaraco', 'David Robertson']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue47060'
versions = ['Python 3.10']
```

</p></details>

**Comment by rgommers** [comment]
I have a very similar case with `importlib.metadata.version` doing something unexpected. Probably the same root cause, but please tell me if it's better to open a separate issue instead.

I built an in-place version of `numpy` with `python setup.py build_ext -i` (note that that will _not_ have a `.dist-info` dir at the root of the repo, `distutils` doesn't produce that), and then used `PYTHONPATH` to switch between the `numpy` installed into the conda env and this in-place build:
```bash
$ echo $PYTHONPATH
/home/rgommers/code/numpy
```
Then open IPython and:
```python
>>> import numpy as np
>>> np.__version__
'1.24.0.dev0+291.g2c5f407cf6'
>>> import importlib.metadata
>>> importlib.metadata.version("numpy")
'1.22.3'
```

So rather than returning `None` or raising an exception, I'm getting the metadata for the wrong package. 

> I haven't yet decided if metadata items being undefined should result in None or raise an Exception (maybe KeyError).

Either way would be fine with me, as long as it's not silently returning incorrect metadata. There will be other cases like that other than `numpy`, and `PYTHONPATH` is pretty commonly used to switch around packages when developing. A `.dist-info` directory may be required to get correct metadata, but it being missing should not cause incorrect results - that's a clear bug.

**Comment by jaraco** [comment]
Hi Ralf. The issue you describe is different, so I'll move it to another bug.

**Comment by sswam** [comment]
FWIW, I noticed that I had two torch directories under `/usr/local/lib/python3.10/dist-packages`:

```
torch-1.13.1+cu116.dist-info  (the installed version)
torch-1.12.1+cu116.dist-info  (containing only an empty file REQUESTED)
```
`importlib_metadata.version("torch")`  was retuning `None` in this scenario. I think it was looking at the directory which only had the REQUESTED file.

After I removed that `torch-1.12.1+cu116.dist-info` directory, it fixed the problem, i.e. now: `importlib_metadata.version("torch") == '1.13.1+cu116'`
